smart-py/marketplace.py

```diff
@@ -1,64 +1,203 @@
+
+        sp.verify(
+            (params.objkt_amount > 0) &
+            (params.objkt_id < self.data.counter) &
+            (params.objkt_id >= 152) &
+            (
+                (params.royalties >= 0) & (params.royalties <= 250)
+            )
+        )
+
+        # TODO
+        # i want to validate the royalties here to make sure they match
+        # what is set on the objkt when it is minted
+
+        # previous for easy comparison
+        # self.fa2_transfer(self.data.objkt, sp.sender, sp.self_address, params.objkt_id, params.objkt_amount)
+
+        self.fa2_transfer(
+            self.data.objkt,
+            sp.sender,
+            sp.self_address,
+            params.objkt_id,
+            params.objkt_amount
+        )
+
+        # previous for easy comparison
+        # self.data.swaps[self.data.counter] = sp.record(issuer=sp.sender, objkt_amount=params.objkt_amount, objkt_id=params.objkt_id, xtz_per_objkt=params.xtz_per_objkt, royalties=params.royalties, creator=params.creator)
+
+        self.data.swaps[self.data.counter] = sp.record(
+            issuer=sp.sender,
+            objkt_amount=params.objkt_amount,
+            objkt_id=params.objkt_id,
+            xtz_per_objkt=params.xtz_per_objkt,
+            royalties=params.royalties,
+            creator=params.creator
+        )
+
         self.data.counter += 1
-    
+
     @sp.entry_point
     def collect(self, params):
         sp.verify(
             # verifies if tez amount is equal to price per objkt
-            (sp.amount == sp.utils.nat_to_mutez(sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some()))) & (self.data.swaps[params.swap_id].objkt_amount != 0))
+            (
+                sp.amount == sp.utils.nat_to_mutez(
+                    sp.fst(
+                        sp.ediv(
+                            self.data.swaps[params.swap_id].xtz_per_objkt,
+                            sp.mutez(1)
+                        ).open_some()
+                    )
+                )
+            # and verify there are objkts available for swap
+            ) & (
+                self.data.swaps[params.swap_id].objkt_amount > 0
+            # and not collecting own swap
+            ) & (
+                sp.sender != self.data.swaps[params.swap_id].issuer
+            )
+        )
 
         sp.if (self.data.swaps[params.swap_id].xtz_per_objkt != sp.tez(0)):
+            self.amount = sp.fst(
+                sp.ediv(
+                    self.data.swaps[params.swap_id].xtz_per_objkt,
+                    sp.mutez(1)
+                ).open_some()
+            )
 
-            self.amount = sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some())
-                
             # calculate fees and royalties
             self.fee = self.amount * (self.data.swaps[params.swap_id].royalties + self.data.fee) / 1000
```

Comment:
Should we store the value of self.data.fee here locally first in case update_fee gets called between this and the next line?

Reply:
That will never happen, i think. From what I understand calls are executed sequentially. You cannot have two concurrent calls at the same time.

Reply:
How is the concurrency defined for these contracts? I'm thinking of a scenario where there could be multiple "threads" and not the same instance calling different methods at the same time.

Reply:
I'm not sure, but I assume that the operations are sorted in some order agreed by the bakers and then they are executed sequentially. Otherwise you could also have problems with multiple collects running at the same time etc. That's why sometimes your collect transactions are rejected: you were unlucky and your position came after other collect operations that collected all the editions.

Reply:
Ah. Ok. Makes sense.

Reply:
love how much learning is going on right now

Reply:
i'd love to hear what @crzypatchwork has to say about this :)